spec/services/token_manager_spec.rb

```diff
@@ -0,0 +1,144 @@
+        expect{subject.issue_token({})}.not_to raise_error
+      end
+    end
+  end
+
+  describe ".secret" do
+    it "uses secret from Rails secret_key_base" do
+      expect(subject.secret).to eq Rails.application.secrets.secret_key_base
+    end
+
+    describe "arguments" do
+      it "accepts only no arguement" do
+        expect{subject.secret}.not_to raise_error
+        expect{subject.secret('arg1', 'arg2')}.to raise_error(ArgumentError, 'wrong number of arguments (2 for 0)')
+        expect{subject.secret('arg1')}.to raise_error(ArgumentError, 'wrong number of arguments (1 for 0)')
+      end
+    end
+  end
+
+  describe ".decode" do
+    let(:token){ subject.generate_token(user.id) }
+
+    it 'decodes information accurately' do
+      decoded = subject.decode(token)
+      expect(decoded.first['user']).to eql user.id
+      expect(decoded.first['exp']).to be_within(100).of(24.hours.from_now.to_i)
+    end
+
+    it 'throws error if token is invalid' do
+      invalid = Faker::Code.ean
+      expect{subject.decode(invalid)}.to raise_error JWT::DecodeError
+    end
+
+    it 'throws error if token is expired' do
+      expired = subject.generate_token(user.id, 1.seconds.from_now)
+      expect{ subject.decode(expired) }.not_to raise_error
+      sleep(1)
+      expect{ subject.decode(expired) }.to raise_error(JWT::ExpiredSignature, 'Signature has expired')
+    end
+
+    describe "arguments" do
+      it "accepts only one arguement" do
+        expect{subject.decode}.to raise_error(ArgumentError, 'wrong number of arguments (0 for 1)')
+        expect{subject.decode({}, {})}.to raise_error(ArgumentError, 'wrong number of arguments (2 for 1)')
+      end
+    end
+  end
+
+  describe ".authenticate" do
+    let(:token){ subject.generate_token(user.id) }
+
+    it 'authenticates a valid token' do
+      decoded = subject.decode(token)
+      expect(decoded.last).to eql('alg' => 'HS512', 'typ' => 'JWT')
+      expect(decoded.first['user']).to eql user.id
+      expect(decoded.first['exp']).to be_within(100).of(24.hours.from_now.to_i)
+    end
+
+    describe "error rescue" do
+      let(:rescued_response){ {} }
+
+      it 'rescues invalid token error' do
+        invalid = Faker::Code.ean
+        expect{subject.decode(invalid)}.to raise_error JWT::DecodeError
+        expect{subject.authenticate(invalid)}.not_to raise_error
+        expect(subject.authenticate(invalid)).to eql rescued_response
+      end
+
+      it 'rescues expired token error' do
+        expired = subject.generate_token(user.id, Time.now)
+        expect{subject.decode(expired)}.to raise_error JWT::ExpiredSignature
+        expect{subject.authenticate(expired)}.not_to raise_error
+        expect(subject.authenticate(expired)).to eql rescued_response
+      end
+    end
+
+    describe "arguments" do
+      it "accepts only one arguement" do
+        expect{subject.authenticate}.to raise_error(ArgumentError, 'wrong number of arguments (0 for 1)')
```

Comment:
I think this may be too rigid, as I can imagine refactoring this sometime later and the `authenticate` method now has to take just one or no parameter.